src/node.h

```diff
@@ -353,18 +354,94 @@ NODE_EXTERN IsolateData* CreateIsolateData(
     ArrayBufferAllocator* allocator = nullptr);
 NODE_EXTERN void FreeIsolateData(IsolateData* isolate_data);
 
-// TODO(addaleax): Add an official variant using STL containers, and move
-// per-Environment options parsing here.
+struct ThreadId {
+  uint64_t id = static_cast<uint64_t>(-1);
+};
+NODE_EXTERN ThreadId AllocateEnvironmentThreadId();
+
+namespace EnvironmentFlags {
+enum Flags : uint64_t {
+  kNoFlags = 0,
+  // Use the default behaviour for Node.js instances.
+  kDefaultFlags = 1 << 0,
+  // Controls whether this Environment is allowed to affect per-process state
+  // (e.g. cwd, process title, uid, etc.).
+  // This is set when using kDefaultFlags.
+  kOwnsProcessState = 1 << 1,
+  // Set if this Environment instance is associated with the global inspector
+  // handling code (i.e. listening on SIGUSR1).
+  // This is set when using kDefaultFlags.
+  kOwnsInspector = 1 << 2,
+  // Set up a standard Node.js Environment, including starting any libuv handles
+  // and the inspector, as well as anything dependent on e.g. environment
+  // variables or command line flags.
+  // This conflicts with LoadEnvironment().
+  // It is recommended to not set or rely on this flag, and it will be removed.
+  kPrepareForExecution = 1 << 3,
+  // Override the --abort-on-uncaught-exception CLI flag to never be active
+  // for this Environment instance.
+  kDoNotAbortOnUncaughtException = 1 << 4
```

Comment:
> However why isn't `kDoNotAbortOnUncaughtException` part of `kOwnsProcessState` for now?

Hm … I’m good with either if you think that makes sense; but I also don’t see any downsides to exposing it this way?

> (It is also possible to modify it on-the-fly and that probably makes a bit more sense) 

I’m honestly not quite sure what you’re referring to? `V8::SetFlagsFromString()`?

> Also why is it `kPrepareForExecution` instead of `kDoNotPrepareForExecution` or `kEnvironmentIndependent` (probably a weird name)?

Because that means that embedders who want to follow the standard startup sequence would have to specify this flag for now, even though it should be removed later – this way, we can just drop it as a default parameter for `CreateEnvironment()` in a semver-major follow-up and basically be done with it.

> As an embedder you'd probably only want to skip the preparation for very rare cases - like for creating an environment-independent snapshot - and if you skip it you only get a half-baked environment e.g. `process` is incomplete, but we have no other public API for them to continue the baking.

I’m not sure if we’re talking past each other, but the preparation should be done as part of `LoadEnvironment()`, which embedders are supposed to call after creating the Environment? It’s just unfortunate that right now we *also* have preparation code in `CreateEnvironment()`…

Reply:
> I’m honestly not quite sure what you’re referring to? V8::SetFlagsFromString()?

I was talking about this, which is consulted when `--should-abort-uncaught-exception` is set:

https://github.com/nodejs/node/blob/2bdeb88c27b4d8de3a8f6b7a438cf0bcb88fa927/src/api/environment.cc#L36-L43

> Because that means that embedders who want to follow the standard startup sequence would have to specify this flag for now

Why isn't this the case for `kDoNotAbortOnUncaughtException`? That's not the default behavior either. I suppose `kDoNotPrepareForExecution` can simply be removed in the semver-major.

> I’m not sure if we’re talking past each other, but the preparation should be done as part of LoadEnvironment(), which embedders are supposed to call after creating the Environment? It’s just unfortunate that right now we also have preparation code in CreateEnvironment()…

Why is that necessary? I think that comes back to the rare cases - the motivation seems to be that we need a half-baked Environment internally, and starting from this PR we use `LoadEnvironment` and  `CreateEnvironment` internally so we are forced to expose this option, but it should be rare for embedders to need a half-baked Environment like us.

Even so I think it's still possible to hide this flag from the public if we simply make a parallel internal enum and cast public flags into it. That way there's no need to make the removal semver-major.

Reply:
> > I’m honestly not quite sure what you’re referring to? V8::SetFlagsFromString()?
> 
> I was talking about this, which is consulted when `--should-abort-uncaught-exception` is set:
> 
> https://github.com/nodejs/node/blob/2bdeb88c27b4d8de3a8f6b7a438cf0bcb88fa927/src/api/environment.cc#L36-L43

I still feel like it could be useful to control this behaviour in a more fine-grained manner by embedders, but I’ve removed the option for now.

> > Because that means that embedders who want to follow the standard startup sequence would have to specify this flag for now
> 
> Why isn't this the case for `kDoNotAbortOnUncaughtException`? That's not the default behavior either. I suppose `kDoNotPrepareForExecution` can simply be removed in the semver-major.

Because `kDoNotAbortOnUncaughtException` was something that was there to give embedders more control over the Environment’s behaviour, whereas `kPrepareForExecution` is there to provide temporary backwards compatibility.

> > I’m not sure if we’re talking past each other, but the preparation should be done as part of LoadEnvironment(), which embedders are supposed to call after creating the Environment? It’s just unfortunate that right now we also have preparation code in CreateEnvironment()…
> 
> Why is that necessary? I think that comes back to the rare cases - the motivation seems to be that we need a half-baked Environment internally, and starting from this PR we use `LoadEnvironment` and `CreateEnvironment` internally so we are forced to expose this option, but it should be rare for embedders to need a half-baked Environment like us.

The current default behaviour of `CreateEnvironment()` is ultimately kind of broken – it creates something that can be used to start a `_third_party_main.js`, but not much more than that. It doesn’t provide anything else, e.g. a way of getting hold of `require()`, which would be necessary if you want anything with more flexibility than `_third_party_main`.

And ultimately, `LoadEnvironment()` is the function that sets up everything that depends on runtime state in some way – that contract is broken by the current state of the API.

> Even so I think it's still possible to hide this flag from the public if we simply make a parallel internal enum and cast public flags into it. That way there's no need to make the removal semver-major.

I don’t quite understand what you mean here, tbh. Stopping `CreateEnvironment()` from running the main thread preparation code is technically semver-major either way, right? 

Or the other way around, how would that internal enum look different from the public one?

Reply:
> The current default behaviour of CreateEnvironment() is ultimately kind of broken – it creates something that can be used to start a _third_party_main.js, but not much more than that. It doesn’t provide anything else, e.g. a way of getting hold of require(), which would be necessary if you want anything with more flexibility than _third_party_main.

How does making it possible to create a half-baked Environment in `CreateEnvironment()` help this? I think having a `LoadEnvironment()` that takes a callback/a main script paired with  `CreateEnvironment()` that creates the entire environment is already enough? `CreateEnvironment()` is not currently where `_third_party_main.js` is started anyway - so I don't quite understand why there is now a comment there saying:

> // TODO(addaleax): This should be removed and CreateEnvironmentInternal() be
>   // renamed to the public API, it's not the right place to run a startup
>   // function here.

What startup function is this referring to?

> And ultimately, LoadEnvironment() is the function that sets up everything that depends on runtime state in some way - that contract is broken by the current state of the API.

I think the fact that `CreateEnvironment()` takes command line arguments kind of implies that this is where the runtime-dependent setups are done? Or at the very least, there is no guarantee that no runtime-dependent setup should be done in `CreateEnvironment()` - the concept that there is such a point where nothing runtime-dependent will be touched in JS is also relatively new.

> Stopping CreateEnvironment() from running the main thread preparation code is technically semver-major either way, right?

I mean if we keep the `kDoNotPrepareExecution`/`kPrepareExecution` option internal and always run the complete JS-land bootstrap in `CreateEnvironment()`, there won't be anything to break.

> Or the other way around, how would that internal enum look different from the public one?

The internal one has `kDoNotPrepareExecution`/`kPrepareExecution` (for the worker code), while the public one does not.